[MRG] clusterQR method added to spectral segmentation

lobpcg opened the pull request:
#### Reference Issues/PRs
closes #12164 

#### What does this implement/fix? Explain your changes.
#12164 adds clusterQR method to 'kmeans' and 'discretize' in spectral clustering

#### Any other comments?
The actual changes for #12164 are just a few lines in only 3 core codes, spectral.py, test_spectral.py, and plot_coin_segmentation.py

lobpcg committed:
clusterQR method added to spectral segmentation

lobpcg committed:
fix comment line too long

lobpcg committed:
typo fixed in spectral.py

lobpcg committed:
spectral.py trailing white space removed

eamanu commented:
@lobpcg Please add [WIP] tag to name

eamanu commented on examples/cluster/plot_coin_segmentation.py:
```diff
@@ -14,12 +14,15 @@
 * with 'kmeans' spectral clustering will cluster samples in the embedding space
   using a kmeans algorithm
+* with 'clusterQR' will cluster samples in the embedding space
+  using a clusterQR algorithm,
```
Please quit the final `,` or add it to the first bullet.

eamanu commented on examples/cluster/plot_coin_segmentation.py:
```diff
@@ -54,28 +57,31 @@
-# Apply spectral clustering (this step goes much faster if you have pyamg
-# installed)
-N_REGIONS = 25
+# the actual number of regions in this example is 27: background and 26 coins
```
Use uppercase in the first letter.

eamanu commented on examples/cluster/plot_coin_segmentation.py:
```diff
@@ -54,28 +57,31 @@
 
 #############################################################################
-# Visualize the resulting regions
+# compute and visualize the resulting regions
```
Use uppercase

eamanu commented on examples/cluster/plot_coin_segmentation.py:
```diff
@@ -54,28 +57,31 @@
+for assign_labels in ('kmeans', 'discretize', 'clusterQR'):
     t0 = time.time()
-    labels = spectral_clustering(graph, n_clusters=N_REGIONS,
+    labels = spectral_clustering(graph, n_clusters=N_REGIONS+N_REGIONS_PLUS,
```
`(N_REGIONS + N_REGIONS_PLUS)` (whitespace and `()`).

eamanu commented:
@lobpcg  look in the `sepectral_clustering` docs.

In the line where Parameters are defined. I think that the Travis error is because the `----` below Parameters title is longer than the title

lobpcg committed:
typos fixed

lobpcg committed:
Update doc/modules/clustering.rst

lobpcg commented:
@eamanu - thanks for your suggestions, now all implemented.

> @lobpcg look in the `sepectral_clustering` docs.
> 
> In the line where Parameters are defined. I think that the Travis error is because the `----` below Parameters title is longer than the title

This was a typo indeed, now fixed - @eamanu thanks for noticing! The fix has however not affected the Travis error, still present unfortunately... Any other advice, please? 

Since the plot_coin_segmentation example demo is involved, I have added the modified fine scikit-learn\doc\modules\clustering.rst to this PR.
I still need to update 
auto_examples/cluster/images/sphx_glr_plot_coin_segmentation_001.png
auto_examples/cluster/images/sphx_glr_plot_coin_segmentation_002.png
and add a new corresponding to clusterQR
auto_examples/cluster/images/sphx_glr_plot_coin_segmentation_003.png

lobpcg committed:
formatting/typo

lobpcg committed:
spectral.py typo fixed

lobpcg commented:
As far as I can see, the PR is ready, other than a single Travis CI error coming from
scikit-learn/sklearn/tests/test_docstring_parameters.py:125: AssertionError
which I need help with fixing, please see above.

eamanu commented:
If you think that is ready please add the tag [MRG]

eamanu commented on sklearn/cluster/spectral.py:
```diff
@@ -42,7 +80,6 @@ def discretize(vectors, copy=True, max_svd_restarts=30, n_iter_max=20,
         Determines random number generation for rotation matrix initialization.
         Use an int to make the randomness deterministic.
         See :term:`Glossary <random_state>`.
-
```
Here is the problem. Add a white line.  This Fix the Travis Error. 

I test it, and this was the problem

lobpcg replied:
Great, thanks very much! Fixed and ready to go.

lobpcg committed:
Update sklearn/cluster/spectral.py

eamanu reviewed (approved):
It's good for me

lobpcg committed:
Merge pull request #1 from scikit-learn/master

master update since Oct 7

lobpcg commented:
This PR has been completed. Could someone please make more comments to react to or just commit it to the base? I start already forgetting my own proposed changes in this PR...

lobpcg commented:
@eamanu This PR has been finalized over a month ago. Could someone please approve it or make new suggestions?

rth commented:
@lobpcg Thanks for the PR, but as was mentioned in the original issue https://github.com/scikit-learn/scikit-learn/issues/12164 by 2 maintainers, this algorithm does not meet the inclusion criteria https://scikit-learn.org/stable/faq.html#what-are-the-inclusion-criteria-for-new-algorithms and would be more suited as a scikit-learn-contrib project until it gains wider usage and could be considered for inclusion in scikit-learn.

lobpcg commented:
The actual changes are just a few lines in only 3 core codes, spectral.py, test_spectral.py, and plot_coin_segmentation.py That hardly justifies, in MHO, creating a brand new separate project at http://contrib.scikit-learn.org/ as proposed by @jnothman and @ogrisel . My hope has been  that @jnothman and @ogrisel might change their recommendation, after seeing the actual changes in the code and results, waiving some of the inclusion criteria due to rather trivial changes in the code and noticeable improvements in performance... I have asked @jnothman and @ogrisel in the original issue #12164 almost 2 months ago, but got no response.

jnothman commented:
> The actual changes are just a few lines in only 3 core codes, spectral.py, test_spectral.py, and plot_coin_segmentation.py That hardly justifies, in MHO, creating a brand new separate project at http://contrib.scikit-learn.org as proposed by @jnothman and @ogrisel .

We simply can't use that justification regularly. Methods for solving optimisation problems change frequently or vary for different applications and we cannot afford to be a complete library of them. This is why we have criteria for maturity. But perhaps if it is highly valuable, we can make spectral segmentation more customisable by allowing a function to be passed instead of one of the built-in solvers?

lobpcg commented:
I surely understand and share your general concerns and the requirements for maturity. However, please consider the following:

1.  I have done all the needed code writing already in this PR, as far as I can see,
2.  the changes are very small, only several lines of code, so there is just tiny additional maintenance effort, 
3.  the new function is not the default in this PR so the code is fully backward compatible, 
4.  the new clusterQR is not yet another optimization method, but rather direct extraction of clusters from eigenvectors in spectral clustering, designed specifically for that purpose only,
5.  in contrast to the existing 'kmeans' and 'discretize', the new clusterQR has no tuning parameters, e.g., runs no iterations,
6.  yet the new clusterQR seems to consistently outperform both 'kmeans' and 'discretize' in terms of quality and speed, e.g., see the example above.

While not technically satisfying yet the criteria for maturity, the new clusterQR appears to set the new standard for extraction of clusters from eigenvectors in spectral clustering, that seems to me to be difficult to improve. It thus deserves in the future to become the default method instead of kmeans after intensive testing by users, that hopefully should follow if this PR is merged and released.

lobpcg committed:
Merge pull request #2 from scikit-learn/master

master merge with destro latest

lobpcg committed:
Merge pull request #3 from lobpcg/master

merge just updated local master into clusterQR

lobpcg commented:
@rh @jnothman @ogrisel I have just merged master into this PR, and it still goes fully green!

jnothman commented on doc/modules/clustering.rst:
```diff
@@ -540,6 +546,14 @@ graph, and SpectralClustering is initialized with `affinity='precomputed'`::
    <http://citeseerx.ist.psu.edu/viewdoc/summary?doi=10.1.1.19.8100>`_
    Andrew Y. Ng, Michael I. Jordan, Yair Weiss, 2001
 
+ * `"Robust and efficient multi-way spectral clustering"
```
not sure that this all needs to be here. it does need to be in the class docstring

lobpcg replied:
I have just added one new reference in the style consistent with all other references around. I have no opinion on appropriateness of the style of the references here. At any rate, I should not be the one changing the style of all these references, if it is justified.

jnothman commented on sklearn/cluster/spectral.py:
```diff
@@ -18,6 +21,41 @@
+    Ut, Vt = svd(vectors[piv, :].T.conj())[0],\
+        svd(vectors[piv, :].T.conj())[2].T.conj()
+    vectors = abs(np.dot(vectors, np.dot(Ut, Vt.T.conj())))
+    return (vectors.argmax(axis=1)).T
```
unnecessary parentheses

lobpcg replied:
done https://github.com/scikit-learn/scikit-learn/pull/12316/commits/ac8445fe98e1c738d781e0aad24f8233a2ab15a7

jnothman commented on sklearn/cluster/spectral.py:
```diff
@@ -18,6 +21,41 @@
+    k = vectors.shape[1]
+    piv = qr(vectors.T.conj(), pivoting=True)[2]
+    piv = piv[0:k]
+    Ut, Vt = svd(vectors[piv, :].T.conj())[0],\
```
should not need `\`. Just assign Ut and Vt in separate statements

jnothman replied:
in fact, why are we computing the SVD twice here?

lobpcg replied:
Great question! I no longer remember and need to re-investigate...

lobpcg replied:
Fixed both, thanks for noticing! The replacement lines are
```
UtSV = svd(vectors[piv, :].T.conj())
    Ut = UtSV[0]
    Vt = UtSV[2].T.conj()
```

jnothman commented on sklearn/cluster/spectral.py:
```diff
@@ -210,10 +248,10 @@ def spectral_clustering(affinity, n_clusters=8, n_components=None,
+        embedding. k-means can be applied and is a popular choice. But it can
         also be sensitive to initialization. Discretization is another
         approach which is less sensitive to random initialization. See
         the 'Multiclass spectral clustering' paper referenced below for
```
this needs to be updated to reference the cluster-qr sources. The references should be added below.

lobpcg replied:
thanks, good catch! Done

lobpcg replied:
all checks passed

jnothman commented on sklearn/cluster/spectral.py:
```diff
@@ -210,10 +248,10 @@ def spectral_clustering(affinity, n_clusters=8, n_components=None,
         when using arpack eigen_solver.
 
-    assign_labels : {'kmeans', 'discretize'}, default: 'kmeans'
+    assign_labels : {'kmeans', 'discretize', 'clusterQR'}, default: 'kmeans'
```
should this just be `qr` rather than `clusterQR`? I think we should avoid caps, and prefer `cluster-qr`.

lobpcg replied:
The original authors call it  `clusterQR` in their paper. It is surely not just `qr`.

I have no personal opinion about the name to use here, and would be happy to rename all instances of  `clusterQR`  into  `cluster-qr` if this is the preferred naming style. Please reconfirm before I make the global change.

lobpcg replied:
@jnothman  - Could you please reconfirm your request before I make the global change?

jnothman reviewed (commented):
I still don't know if this is something we want, but it would be good if it had some more tests of expected behaviour

lobpcg committed:
github.com/scikit-learn/scikit-learn/pull/12316#discussion_r251665191

lobpcg committed:
https://github.com/scikit-learn/scikit-learn/pull/12316#discussion_r251665390

lobpcg commented:
Some good examples would be https://github.com/scikit-learn/scikit-learn/issues/12235 but this is a bit of work and unclear if there is enough support for it

lobpcg committed:
Merge pull request #5 from scikit-learn/master

master update with the latest

lobpcg committed:
Merge pull request #6 from lobpcg/master

update clusterQR with the latest master

lobpcg commented:
Trying to run the code to look at the issues raised by @jnothman , I just run 
scikit-learn\examples\cluster\plot_coin_segmentation.py 
after pooling the latest master, but I now get multiple pylint errors, coming from 
\scikit-learn\sklearn\preprocessing\data.py 
e.g., from line 33:

No module named 'sklearn.preprocessing._csr_polynomial_expansion'

I do not understand why. May be I have forgotten how, e.g. running examples requires some preliminary steps? Could someone please advise?

lobpcg committed:
Merge pull request #8 from scikit-learn/master

master update

lobpcg committed:
Merge pull request #9 from lobpcg/master

update from latest master

lobpcg committed:
Update spectral.py

fix for https://github.com/scikit-learn/scikit-learn/pull/12316#discussion_r251665374

lobpcg committed:
removed redundant SVD

lobpcg committed:
testing new tol 1e-5 defaults and the laplacian shift in AMG

lobpcg commented:
I have also applied the fix for #13393, by adding a line to AMG 
```
  laplacian = laplacian + 1e-10 * sparse.eye(laplacian.shape[0])
'''
and lowering the default tolerance to 1e-5 in AMG and LOBPCG in sklearn/manifold/spectral_embedding_.py so that any solver now works in plot_coin_segmentation.py

lobpcg commented:
@jnothman Could you please have another look? I've made all the self-evident changes you request and additionally applied a new fix for  #13393. 

I am unsure about your suggestion to rename clusterQR into cluster-qr.

lobpcg committed:
changed to eigen_solver='amg'

lobpcg committed:
eigen_solver='amg' is not installed in CircleCI  so back to 'arpack'

lobpcg committed:
comment updated to include amg and lobpcg, now fixed

whitews commented:
Any update on this PR? I've run into the AMG bug and was hoping the fix could be included in 0.20.4.

jnothman commented:
Do you mind splitting this to separate out the fix for AMG? I'd like to see it on its own with a non-regression test.

whitews commented:
@jnothman  Are you asking me or @lobpcg ? Andrew, I'm happy to do it if you don't have the time.

lobpcg commented:
My proposed fixes for both reference Issues/PRs #12164<https://github.com/scikit-learn/scikit-learn/issues/12164> and #13393 <https://github.com/scikit-learn/scikit-learn/issues/13393> are very small and interrelated though scikit-learn\examples\cluster\plot_segmentation_toy.py
I would of course prefer that this PR gets enough attention as is and finally is merged so that I could move on to other changes I am interested to propose. To split the two just adds more work, probably unnecessary since the changes in the code are actually very small in both issues.
On 4/3/2019 8:47 PM, Scott White wrote:

@jnothman<https://github.com/jnothman> Are you asking me or @lobpcg<https://github.com/lobpcg> ? Andrew, I'm happy to do it if you don't have the time.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/scikit-learn/scikit-learn/pull/12316#issuecomment-479708272>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AorJvbrY6VAMIiMs0vP85zhMWG_fj4Lrks5vdUuZgaJpZM4XLbha>.

jnothman commented on sklearn/manifold/spectral_embedding_.py:
```diff
@@ -288,11 +288,12 @@ def spectral_embedding(adjacency, n_components=8, eigen_solver=None,
         laplacian = check_array(laplacian, dtype=np.float64,
                                 accept_sparse=True)
         laplacian = _set_diag(laplacian, 1, norm_laplacian)
+        laplacian = laplacian + 1e-5 * sparse.eye(laplacian.shape[0])
```
Okay. Then are you able to directly test these changes?

lobpcg replied:
Done; see https://github.com/scikit-learn/scikit-learn/pull/12316#issuecomment-481841935

lobpcg commented:
Yes, please see

https://github.com/scikit-learn/scikit-learn/pull/12316/commits/a11d94a754aee6bd4619477e9b424eb03eba30ba

But eigen_solver='amg' is not installed in CircleCI so fails the tests, thus the default cannot be changed, see

https://github.com/scikit-learn/scikit-learn/pull/12316/commits/e1fd3a058e4822e6bbff9defb393a0f5f2842e12

One can write a separate new test, specifically for this fix - may be @whitews is interested to do it? I am actually off-line skiing in CO this week :-)

The simplest Laplacian to test on could be a 1D Laplacian with periodic boundary conditions, e.g., generated by [lambda,V,A] = laplacian([100],{'P'}, 2); from https://www.mathworks.com/matlabcentral/fileexchange/27279-laplacian-in-1d-2d-or-3d Something similar, but with periodic boundary conditions, to https://somethinglikethis577.wordpress.com/2017/05/18/laplace-in-1d-in-python/

On 4/4/2019 1:04 AM, Joel Nothman wrote:

@jnothman commented on this pull request.

________________________________

In sklearn/manifold/spectral_embedding_.py<https://github.com/scikit-learn/scikit-learn/pull/12316#discussion_r272018164>:

> @@ -288,11 +288,12 @@ def spectral_embedding(adjacency, n_components=8, eigen_solver=None,
         laplacian = check_array(laplacian, dtype=np.float64,
                                 accept_sparse=True)
         laplacian = _set_diag(laplacian, 1, norm_laplacian)
+        laplacian = laplacian + 1e-5 * sparse.eye(laplacian.shape[0])


Okay. Then are you able to directly test these changes?

jnothman commented:
A unit test, rather than the example would be the goal. Examples do
implicitly test some features, but they're not really there for quality
assurance.

lobpcg commented:
Where this test should go?  AMG is not installed in CircleCI so it would fail if added to standard tests.

A unit test, rather than the example would be the goal. Examples do
implicitly test some features, but they're not really there for quality
assurance.

jnothman commented:
unit test in `sklearn/manifold/tests/test_spectral_embedding.py`, please. Other tests there require AMG, and this is tested on one of our Azure instances.

lobpcg commented:
@jnothman in order to consistently break AMG in  `sklearn/manifold/tests/test_spectral_embedding.py` I had to modify the test case to make the graph not fully connected, so it now generates the following warning 

> Graph is not fully connected, spectral embedding may not work as expected.

 before LOBPCG fails. I am actually unsure why this warning about the fully connected graph is here, since the spectral embedding should work either way to my knowledge. Anyway, with the proposed fix  `laplacian = laplacian + 1e-5 * sparse.eye(laplacian.shape[0])` the AMG code no longer fails in the modified test. AMG is also tested in `sklearn\cluster\tests\test_spectral.py`, where my fix passes as well. 

I'll commit the change now.

lobpcg committed:
AMG test on NOT fully connected graph

lobpcg committed:
last changes reversed for error debugging

lobpcg commented:
Could someone please advise? I've reversed my today changes, but some of the tests still fail with funny error messages. My previous commit about a month ago has passed all tests.

lobpcg committed:
new AMG example, still connected graph

lobpcg committed:
change n_neighbors=5 to 3 for graph to disconnect

lobpcg committed:
reversed changes to get back the original

lobpcg committed:
added new AMG test NOT fully connected graph

lobpcg committed:
flake8 line under-indented fixes

lobpcg commented:
CircleCI tests are clearly misbehaving now and thus ignored

lobpcg committed:
Merge pull request #11 from scikit-learn/master

master update

lobpcg committed:
Merge pull request #12 from lobpcg/master

clusterQR update from the latest master

lobpcg commented:
@jnothman @ogrisel @eamanu  @whitews 
 The PR is finished and passes all tests on the latest master. Could it please be included in 0.20.4?

jnothman commented:
I'm not sure of 0.20.4 will happen (ever), and this could not be merged into a bug fix release as it adds new API. We are hoping to release 0.21 in the next couple of weeks, and this will probably miss the boat, because it would require two full reviews. I had suggested splitting this to separate the bug fix from the enhancement / new feature such that we could include at least the bug fix in an upcoming release without necessarily committing to clusterQR support.

jnothman commented:
> I am actually unsure why this warning about the fully connected graph is here, since the spectral embedding should work either way to my knowledge.

In the PR where that warning about multiple connected components first appeared, there were anecdotal comments about it working poorly from @weilinear and @GaelVaroquaux, in a discussion around 2 Oct 2012.

> Anyway, with the proposed fix laplacian = laplacian + 1e-5 * sparse.eye(laplacian.shape[0]) the AMG code no longer fails in the modified test.

Yes, I agree the test seems effective. If you don't mind us pulling that fix out into a separate PR, we could squeeze it into 0.21 where the rest of the PR here is unlikely to make it.

lobpcg commented:
> > I am actually unsure why this warning about the fully connected graph is here, since the spectral embedding should work either way to my knowledge.
> 
> In the PR where that warning about multiple connected components first appeared, there were anecdotal comments about it working poorly from @weilinear and @GaelVaroquaux, in a discussion around 2 Oct 2012.

I suspect that the actual problem is not with fully connected or not graphs in the spectral embedding, but rather with the incorrect use/call of the eigenvalue solvers, just as in this fix for AMG/LOBPCG.  While playing with changing n_samples in the sklearn\cluster\tests\test_spectral.py I notice that several tests may fail for different values of n_samples for really no good reason. I thus suspect that there might be other troubles that require investigating, e.g., with calling ARPACK eigs. 

> > Anyway, with the proposed fix laplacian = laplacian + 1e-5 * sparse.eye(laplacian.shape[0]) the AMG code no longer fails in the modified test.
> 
> Yes, I agree the test seems effective. If you don't mind us pulling that fix out into a separate PR, we could squeeze it into 0.21 where the rest of the PR here is unlikely to make it.

Sure, please go ahead. I've started doing it myself but ran out of time to finish. Please add the references given in https://github.com/scikit-learn/scikit-learn/issues/13393 to the scikit-learn docs. May be state explicitely in the docs that AMG is actually LOBPCG with AMG preconditioning - the current AMG description in scikit-learn docs is incomplete and misleading, making one (including me originally) to think that it's stand-alone AMG solver for eigenproblems.

whitews commented:
@lobpcg  I have re-implemented the AMG fix you proposed but am still getting LinAlgError that the "array is not positive definite" for some seeds. I'm using the small example script you provided in issue #13393. Any ideas?

lobpcg commented:
> @lobpcg I have re-implemented the AMG fix you proposed but am still getting LinAlgError that the "array is not positive definite" for some seeds. I'm using the small example script you provided in issue #13393. Any ideas?

The code of LOBPCG itself went though multiple fixes recently in SciPy 1.2.1 and more coming in 1.3.0, see https://github.com/scipy/scipy/pull/9650   What is your SciPy version?

whitews commented:
Ok, looks like I may have had an older SciPy version, created a new virtual environment with 1.2.1 and cannot reproduce the problem. I'll test it some more with my own code where I first encountered the issue, then submit the PR. Seems it works well with either:

`laplacian = laplacian + 1e-10 * sparse.eye(laplacian.shape[0])`

or 

`laplacian = laplacian + 1e-5 * sparse.eye(laplacian.shape[0])`

Do you still recommend 1e-5? I believe you mentioned somewhere that it may be safer for single precision.

lobpcg commented:
From #13393: 
> The choice of alpha=1e-10 here should be OK in double-precision, but I would advocate alpha=1e-5 to be also safe in single-precision. Choosing an increasingly positive alpha may be expected to slow down the LOBPCG convergence, e.g., alpha=1 is probably excessively large, while the change from alpha=1e-10 to alpha=1e-5 would probably be unnoticeable.

LOBPCG has not been tested much in single precision even without AMG and may experience numerical troubles by itself, so it is better be on the safe side.

lobpcg commented:
My fix for  #13393 (makes AMG to work without failure) is attempted  in https://github.com/scikit-learn/scikit-learn/pull/13707

If it's merged, I will need to revert all the changes here, going back to 
https://github.com/scikit-learn/scikit-learn/pull/12316/commits/6b00bbbdc0742fc43b1259558c058688f9f49d6c

lobpcg committed:
Merge pull request #14 from scikit-learn/master

master update

lobpcg committed:
Merge branch 'clusterQR' into master

lobpcg committed:
Merge pull request #16 from lobpcg/master

clusterQR update

lobpcg commented:
All, @jnothman @eamanu @whitews   I have merged with the latest master, and there are still no errors/conflicts. This is my last activity in this PR. 

@whitews it would be nice if after our #13707 is finally merged, you could get back here to revert all the AMG changes, back to pre-AMG 6b00bbb

Writing and then maintaining this PR has taken lots time, and was fun. But I am no longer motivated and have no time to work on this. If anyone wants to take over, please let me know - I'd still be able to help with an advice. If not, this effort will naturally age and die.

lobpcg committed:
Merge branch 'master' into clusterQR

lobpcg committed:
reverse changes that appear in a separate PR

lobpcg committed:
reverse changes that appear in a separate PR

lobpcg committed:
remove conj - complex numbers are not neeed

lobpcg committed:
reverse irrelevant change

lobpcg committed:
reverse changes not relevant to this PR

lobpcg committed:
reverse irrelevant changes to this PR

lobpcg committed:
typo fixed

lobpcg committed:
reverse changes irrelevant to this PR

lobpcg committed:
Merge pull request #18 from scikit-learn/master

master update

lobpcg commented:
@GaelVaroquaux could you please have a look at this PR and express your opinion?

ogrisel commented:
The AMG fix #13707 is now merged in master.

lobpcg committed:
Merge branch 'master' into clusterQR

lobpcg committed:
Merge pull request #21 from lobpcg/clusterQR

Cluster qr

lobpcg committed:
Merge pull request #22 from lobpcg/master

Merge pull request #21 from lobpcg/clusterQR

lobpcg commented:
Could someone please review and finally decide if this dies or gets merged?

lobpcg commented:
Could someone please review?

agramfort commented:
in the referenced issue I see:

```
This doesn't meet our basic criteria for inclusion of stable and mature algorithms. What makes you think it is worth our while to maintain an implementation of this? What are the chances that this will remain a canonical approach in 5 years' time?
```
 
```
+1 for making a prototype Python implementation outside of the scikit-learn code base and running some benchmarks.
```

so there are some concerns which could eventually be ruled out with some benchmarks showing the empirical superiority of the method. Have these benchmarks done and made public?

lobpcg commented:
> 
> 
> in the referenced issue I see:
> 
> ```
> This doesn't meet our basic criteria for inclusion of stable and mature algorithms. What makes you think it is worth our while to maintain an implementation of this? What are the chances that this will remain a canonical approach in 5 years' time?
> ```
> 
> ```
> +1 for making a prototype Python implementation outside of the scikit-learn code base and running some benchmarks.
> ```
> 
> so there are some concerns which could eventually be ruled out with some benchmarks showing the empirical superiority of the method. Have these benchmarks done and made public?

@agramfort Yes, of course, please see the plots in https://github.com/scikit-learn/scikit-learn/pull/12316#issuecomment-427611065 and/or run the plot_coin_segmentation example demo in this PR yourself. A systematic comparison is performed in the original paper, cited in https://github.com/scikit-learn/scikit-learn/issues/12164

agramfort commented:
>
> the plot_coin_segmentation example demo in this PR
>

tells me the code leads to comparable results.

> A systematic comparison is performed in the original paper, cited in
> #12164 <https://github.com/scikit-learn/scikit-learn/issues/12164>
>
have a script to replicate the timings given the current implementation is
necessary.

see for example the benchmarks
https://github.com/scikit-learn/scikit-learn/tree/master/benchmarks
that have been shared with the PRs to demonstrate and replicate timings on
different
problem dimensions and hardware.

lobpcg commented:
> the plot_coin_segmentation example demo in this PR
> tells me the code leads to comparable results.

Comparable, but better - please just look at the plots. 

> have a script to replicate the timings given the current implementation is necessary. see for example the benchmarks https://github.com/scikit-learn/scikit-learn/tree/master/benchmarks that have been shared with the PRs to demonstrate and replicate timings on different problem dimensions and hardware.

Sure, if this is a blocker, it is doable and probably not too difficult. I'll look into it - thanks for the link!

lobpcg commented:
@ glemaitre  - could you please review?

Also, there is now a conflict with sklearn/cluster/spectral.py - could someone please resolve it?
My "resolve" button is greyed out somehow

lobpcg committed:
trying to fix a conflict

lobpcg committed:
Merge pull request #23 from scikit-learn/master

master update

lobpcg committed:
Merge pull request #24 from lobpcg/master

update from latest master

lobpcg committed:
restore the edits

lobpcg committed:
add imports

lobpcg committed:
trying to fix rst warnings

lobpcg commented:
This PR is stalled and has conflicts. I close it and open instead [WIP] cluster_qr method added to spectral segmentation #21148
